common/transport/amqp/src/amqp.ts

```diff
@@ -135,6 +88,280 @@ export class Amqp {
         forever: false
       }
     }, amqp10.Policy.EventHub));
+
+    this._fsm = new machina.Fsm({
+      namespace: sdkVersionString,
+      initialState: 'disconnected',
+      states: {
+        disconnected: {
+          _onEnter: (disconnectCallback, err, result) => {
+            if (err) {
+              this._safeCallback(disconnectCallback, err);
+            } else {
+              this._safeCallback(disconnectCallback, null, new results.Disconnected());
+            }
+          },
+          amqpError: (err, callback) => { callback(); }, // we're already considering ourselves disconnected. it'd be a bug though?
```

Comment:
>it'd be a bug though? [](start = 110, length = 21)

This could either be a race condition (error arrives right after the disconnect completed), or a big screw up (lower levels are in a different state than upper levels think they are.  Code typically can't recover from this very nicely if we're not careful))

Reply:
yup, I don't want to crash though, but I'll add a debug log line

---
In reply to: [139507770](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139507770) [](ancestors = 139507770)

Reply:
Fair enough.

---
In reply to: [139538015](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139538015) [](ancestors = 139538015,139507770)